src/core/lwt_stream.ml

```diff
@@ -130,45 +129,76 @@ let clone s =
      | _ -> ());
   {
     source = s.source;
+    close = s.close;
     node = s.node;
     last = s.last;
-    hooks = s.hooks;
   }
 
-let from f =
+let from_source source =
   let last = new_node () in
-  {
-    source = From { from_create = f; from_thread = Lwt.return_unit };
-    node = last;
-    last = ref last;
-    hooks = ref [];
+  let _, close = Lwt.wait () in
+  { source = source
+  ; close = close
+  ; node = last
+  ; last = ref last
   }
 
+let from f =
+  from_source (From { from_create = f; from_thread = Lwt.return_unit })
+
 let from_direct f =
-  let last = new_node () in
-  {
-    source = From_direct f;
-    node = last;
-    last = ref last;
-    hooks = ref [];
-  }
+  from_source (From_direct f)
+
+let closed s =
+  Lwt.waiter_of_wakener s.close
+
+let is_closed s =
+  not (Lwt.is_sleeping (closed s))
 
 let on_termination s f =
-  s.hooks := f :: !(s.hooks)
+  Lwt.async (fun () -> closed s >|= f)
 
 let on_terminate = on_termination
 
+let enqueue' e last =
+  let node = !last and new_last = new_node () in
+  node.data <- e;
+  node.next <- new_last;
+  last := new_last
+;;
+
+let enqueue e s =
+  enqueue' e s.last
+
+let feed_direct f last close  =
+  let x = f () in
+  (* Push the element to the end of the queue. *)
+  enqueue' x last;
```

Comment:
@aantron this is pushing elements onto the queue. `feed_direct` is called in a loop in `from_fixed`.